### PR TITLE
Add PCAttentionCoupleBatchNegative

### PR DESCRIPTION
**Pull request by pamparamm**
This PR adds a new node: `PCAttentionCoupleBatchNegative` - it sets "hooks" in `negative` cond to HookGroup from `positive` cond by reference **if and only if** their `hooks` lists are identical by reference, causing `cond` and `uncond` to be placed in the same`hooked_to_run` record.
Using this node, attention couple from Schedule Prompt achieves the same speed as AttentionCouplePPM (and also generates the same exact images).

I've also reverted some incompatible optimizations (and `pop` side-effects) in `AttentionCoupleHook` class.

Related to #108

### CODE REVIEW

**Comment by pamparamm** [comment]
> While you're here, do you have any ideas on how we could avoid your nodes having to reimplement and monkeypatch the advanced encoding functions and still maintain NegPiP support? I'm thinking of cleaning up the implementation a bit (I already did some work to fix non-CLIP encoders)

I think `encode_regions` from `cutoff` can be replaced with my `encode_regions_negpip` version without any issues, since the compatibility logic is inside two simple if-checks.

As for advanced encoding patch - I need to think, it's very hacky (but also quite simple).

**Comment by asagi4** [comment]
Though that pop might cause issues if the hook is used elsewhere without copying the base cond's setting dictionary first.

**Comment by asagi4** [comment]
This needs a bit more testing with crazier cond setups, but as long as the batch negative node falls back to the slower mode of operation when it can't work, I think it's enough of an improvement as is. I guess if you wanted to support attention masks for negative conds with batching, it should be possible if you extend the attention mask hook with an optional "negative_masks" parameter that the BatchNegative node sets.

In the case the negative cond has a hook on it, the node would in essence do something like the following:
``` 
combined_hook = NewHook()
combined_hook.masks = pos.hook.masks
combined_hook.neg_masks = neg.hook.masks
pos.hook = neg.hook = combined_hook 
```
and then you'd have some extra logic in the hook to handle the case where neg_masks set.

**Comment by asagi4** [comment]
... I tried your node with one of my comfybox workflows and for some reason *adding it causes both conds to become None apparently* ???

It's not your code; ComfyUI isn't even calling the node. But something about it is triggering a misbehaviour somewhere.

How am I hitting all these bugs...

File: prompt_control/nodes_hooks.py [hunk 1]
```diff
@@ -79,10 +84,51 @@ def create_hook(loraspec, start_pct, end_pct, non_scheduled):
...
+            log.warning("Batching scheduled negatives is not supported yet")
+            return (positive, negative)
+
+        negative_batch = []
+        for p in positive:
+            n = [negative[0][0], negative[0][1].copy()]
+            n_hook_group: comfy.hooks.HookGroup = n[1].get("hooks", comfy.hooks.HookGroup()).clone()
+            p_hook_group: comfy.hooks.HookGroup = p[1].get("hooks", comfy.hooks.HookGroup())
+            attn_couple = [hook for hook in p_hook_group.hooks if isinstance(hook, AttentionCoupleHook)]
+            for hook in attn_couple:
```
Comment (asagi4) [inline prompt_control/nodes_hooks.py:116]: I think n_hook_group needs to be cloned before you modify it here, otherwise you will mutate the input.

**Comment by asagi4** [comment]
I think there are some cases with scheduling where the batching breaks if the scheduled prompts have different sizes.
For example, something like this seems to cause a tensor size error:
```
a a a  a a a a a a a  a a a a a a a  a a a a a a a  a a a a a a a  a a a a a a a  a a a a a a a  a a a a a a a  a a a a a a a  a a a a a a a  a a a a a a a  a a a a a a a  a a a a a a a  a a a a a a a  a a a a a a a  a [a a a a a a  a a a a a a a  a a a a a a a  a a a a a a a  a a a a a ATTN() AND ATTN() a a  a a a a a a a  a a a a a a a  a a a a a a a  a a a a a a a  a a a a a a a  a a a a a a a  a a a a a a a  a a a a a a a  a a a a a a a  a a a a a a a  a a a a a a a  a a a a a a a  a a a a a a a  a a a a a a a  a a a a a a a  a a a a a a a  a a a a a a a  a a a a a a a  a a a a a a a  a:0.1] a a a a a a  a a a a a a a  a a a a a a a  a a a a a a a  a a a a a a a  a a a a a a a  a a a a a a a  a a a a 
```
negative prompt empty.

**Comment by asagi4** [comment]
I think it might break in cases where LCM of (negtokens+postokens) is different from just LCM(postokens)?

**Comment by asagi4** [comment]
Except something has broken with batch size 2 now... of course
